### PR TITLE
Feature/interface to erforce simulator

### PR DESCRIPTION
**Pull request by emielsteerneman**
Added code so that simulator commands (add/remove robots, teleport ball, etc) can be send from the interface to a simulator via AI

### CODE REVIEW

File: roboteam_ai/src/interface_api/InterfaceSubscriber.cpp [hunk 1]
```diff
@@ -47,21 +50,27 @@ void InterfaceSubscriber::onMessage(const proto::MsgFromInterface&& message) {
...
             RTT_ERROR("Received message with no kind set");
             break;
     }
 }
 
+void InterfaceSubscriber::sendPacketToSimulator(const SimulatorCommand& packet) {
+    QByteArray datagram;
+    datagram.resize(packet.ByteSizeLong());
+    packet.SerializeToArray(datagram.data(), datagram.size());
+    auto bytesSent = simulator_socket.writeDatagram(datagram, QHostAddress::LocalHost, 10300);
```
Comment (mamiksik) [inline roboteam_ai/src/interface_api/InterfaceSubscriber.cpp:73]: I don't like that the UI subscriber has dependency on QT. What about moving this to IOManager and just invoking it in the switch here?

File: roboteam_ai/include/roboteam_ai/interface_api/InterfaceSubscriber.h [hunk 2]
```diff
@@ -6,11 +6,11 @@
 #define RTT_INTERFACESUBSCRIBER_H
 #include <atomic>
 #include "proto/NewInterface.pb.h"
+#include <QtNetwork>
 
 namespace rtt::ai::io {
 class InterfaceGateway;
 class InterfaceSubscriber {
-    friend InterfaceGateway;  /// Only the InterfaceGateway can create an InterfacePublisher
```
Comment (mamiksik) [inline roboteam_ai/include/roboteam_ai/interface_api/InterfaceSubscriber.h:13]: Why did you remove this and made the constructor public?

File: roboteam_interface_vue/src/modules/field/game-canvas.vue [hunk 3]
```diff
@@ -9,7 +9,7 @@ import Robots from './robots-graphics.vue'
 import Ball from './ball-graphics.vue'
 import Field from './field-graphics.vue'
 import Visualizations from './pointer-location.vue'
-import PointerLocation from './visualizarions-graphics.vue'
```
Comment (mamiksik) [inline roboteam_interface_vue/src/modules/field/game-canvas.vue:12]: haha ty for fixing typos!

File: roboteam_interface_vue/src/modules/composables/ai-controller.ts [hunk 4]
```diff
@@ -29,6 +30,43 @@ export const useAiController = defineStore('aiController', () => {
...
+      'teleportBall' : {
+        'x' : 0, 'y' : 0, 'z' : 0, 'vx' : 0, 'vy' : 0, 'vz' : 0
+      }
+    }
+  }
+
+  const commandMap = new Map<string, Object>([
+    ['robotsToSide', robotsToSide],
+    ['ballToCenter', ballToCenter],
+  ])
```
Comment (mamiksik) [inline roboteam_interface_vue/src/modules/composables/ai-controller.ts:64]: I think this would be better placed directly in the simulator-widget. I cannot think of another place where would you invoke this and, thus, as is it pollutes the ai-controller

File: roboteam_interface_vue/src/modules/composables/ai-controller.ts [hunk 5]
```diff
@@ -29,6 +30,43 @@ export const useAiController = defineStore('aiController', () => {
...
+
+  const sendSimulatorCommand = (command: string) => {
+    send('simulatorCommand', commandMap.get(command))
+  }
```
Comment (mamiksik) [inline roboteam_interface_vue/src/modules/composables/ai-controller.ts:68]: This I would keep there, but then instead of passing command as string I would pass the command object

Comment (mamiksik) [inline roboteam_interface_vue/src/modules/composables/ai-controller.ts:68]: also there is a shorthand syntax ` const sendSimulatorCommand = (command: Object) => send('simulatorCommand', commnad);`

btw: isn't there a type for the proto packet? So it's not just annotated `Object`?

**Review by mamiksik — CHANGES_REQUESTED** [review]
Hey, good job but there are few things that need to be fixed. 

Btw: what about adding event hendler for right click, to teleport selected robots to cursor location?

File: roboteam_ai/include/roboteam_ai/interface_api/InterfaceSubscriber.h (hunk 2 above)
Comment (emielsteerneman) [reply roboteam_ai/include/roboteam_ai/interface_api/InterfaceSubscriber.h:13]: The `InterfaceSubscriber` is now holding a QUdpSocket, which means that it needs to be instantiated. Since the `InterfaceSubscriber` is wrapped in a `std::unique_ptr`, it can only be instantiated if the constructor is public

File: roboteam_ai/src/interface_api/InterfaceSubscriber.cpp (hunk 1 above)
Comment (emielsteerneman) [reply roboteam_ai/src/interface_api/InterfaceSubscriber.cpp]: Done

File: roboteam_interface_vue/src/modules/composables/ai-controller.ts (hunk 5 above)
Comment (emielsteerneman) [reply roboteam_interface_vue/src/modules/composables/ai-controller.ts:68]: Done

File: roboteam_interface_vue/src/modules/composables/ai-controller.ts (hunk 4 above)
Comment (emielsteerneman) [reply roboteam_interface_vue/src/modules/composables/ai-controller.ts:64]: Done

**Review by mamiksik — APPROVED** [review]
There are things to fix/add for sure but imho it's a good starting point thus I give it LGTM

**Review by Eems101 — APPROVED** [review]
